x/zkism/keeper/abci.go

```diff
@@ -0,0 +1,54 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+
+	"cosmossdk.io/collections"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// BeginBlocker contains the implementation of the appmodule.BeginBlock interface.
+// This is an ABCI lifecycle method and is called at the beginning of block finalization through
+// ABCI's FinalizeBlock method.
+// BeginBlocker is used to track header hashes
+func (k *Keeper) BeginBlocker(ctx context.Context) error {
+	return k.StoreHeaderHash(ctx)
+}
+
+// StoreHeaderHash stores the header hash keyed by block height and prunes the oldest entry
+func (k *Keeper) StoreHeaderHash(goCtx context.Context) error {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	numEntries, err := k.GetMaxHeaderHashes(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Prune store to ensure we only have parameter-defined entries.
+	// In most cases, this will involve removing a single entry.
+	// In the rare scenario when the entries gets reduced to a lower value k'
+	// from the original value k. k - k' entries must be deleted from the store.
+	// Since the entries to be deleted are always in a continuous range, we can iterate
+	// over the historical entries starting from the most recent version to be pruned
+	// and then return at the first empty entry.
+	for i := ctx.BlockHeight() - int64(numEntries); i >= 0; i-- {
```

Comment:
should the numEntries be limited? To avoid having a very long loop?

Reply:
The loop counter is decrementing, and the loop will `break` below when we lookup and get `collections.ErrNotFound`.

We start at current height - `numEntries` so that we always retain `numEntries` worth of header hashes. 

E.g. imagine numEntires is 10000. The current height 10001, we start at 1, lookup and prune at height 1, and then add 10001 below on L53.

As mentioned in the godoc, in most cases we will always be pruning a single entry and then breaking on the next iteration of the loop. Unless the `numEntires` has been lowered, in which case we need to prune `k - k'`

Reply:
thanks